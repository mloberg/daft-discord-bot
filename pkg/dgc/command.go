package dgc

import "github.com/bwmarrin/discordgo"

// Command represents a Discord slash command
type Command struct {
	Name        string
	Description string
	Options     []*discordgo.ApplicationCommandOption

	Run func(s *discordgo.Session, i *discordgo.InteractionCreate) error
}
