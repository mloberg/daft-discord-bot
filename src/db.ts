import { PrismaClient } from '@prisma/client';

const client = new PrismaClient({
    // TODO: setup logging
});

export default client;
