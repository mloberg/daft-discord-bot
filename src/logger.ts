import fs from 'fs';
import pino from 'pino';

import config from './config';

const logger = pino(
    {
        level: config.logLevel,
        prettyPrint: config.debug,
    },
    config.isTest ? fs.createWriteStream('/dev/null') : process.stderr,
);

export default logger;
