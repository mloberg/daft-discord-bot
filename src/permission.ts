import { Guild, GuildMember } from 'discord.js';

import config from './config';

export const ensureRole = async (guild: Guild): Promise<void> => {
    const role = guild.roles.cache.find((role) => role.name === config.djRole);
    if (role) {
        return;
    }

    await guild.roles.create({ data: { name: config.djRole } });
};
