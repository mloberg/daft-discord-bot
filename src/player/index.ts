import { StreamDispatcher, StreamOptions, VoiceConnection } from 'discord.js';

import { Player as PlayerInterface } from '../types';
import LocalFile from './local';

export class Player implements PlayerInterface {
    private players: PlayerInterface[];

    constructor(...players: PlayerInterface[]) {
        this.players = players;
    }

    play(location: string, connection: VoiceConnection, options: StreamOptions = {}): Promise<StreamDispatcher> {
        const player = this.players.find((player) => player.supports(location));
        if (!player) {
            throw new Error(`Could not find player for ${location}`);
        }

        return player.play(location, connection, options);
    }

    getTitle(location: string): Promise<string | null> {
        const player = this.players.find((player) => player.supports(location));
        if (!player) {
            throw new Error(`Could not find player for ${location}`);
        }

        return player.getTitle(location);
    }

    supports(location: string): boolean {
        for (const player of this.players) {
            if (player.supports(location)) {
                return true;
            }
        }

        return false;
    }
}

export default new Player(new LocalFile());
