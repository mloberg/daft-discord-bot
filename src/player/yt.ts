import { StreamDispatcher, StreamOptions, VoiceConnection } from 'discord.js';
import ytdl from 'ytdl-core-discord';

import { Player } from '../types';

export default class YouTube implements Player {
    async play(location: string, connection: VoiceConnection, options: StreamOptions = {}): Promise<StreamDispatcher> {
        options.type = 'opus';
        return connection.play(await ytdl(location, { highWaterMark: 1 << 25 }), options);
    }

    async getTitle(location: string): Promise<string | null> {
        const info = await ytdl.getBasicInfo(location);
        return info.title;
    }

    supports(location: string): boolean {
        return ytdl.validateURL(location);
    }
}
