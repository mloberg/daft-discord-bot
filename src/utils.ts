export function env(key: string, _default: string): string {
    return process.env[key] ?? _default;
}
