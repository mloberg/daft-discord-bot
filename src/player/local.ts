import { exec } from 'child_process';
import { StreamDispatcher, StreamOptions, VoiceConnection } from 'discord.js';
import { createReadStream, existsSync } from 'fs';
import { extname } from 'path';
import { promisify } from 'util';

import logger from '../logger';
import { Player } from '../types';

const shell = promisify(exec);

export default class LocalFile implements Player {
    async play(location: string, connection: VoiceConnection, options: StreamOptions = {}): Promise<StreamDispatcher> {
        const ext = extname(location);

        if ('.ogg' === ext) {
            return connection.play(createReadStream(location), { ...options, type: 'ogg/opus' });
        }

        if ('.webm' === ext) {
            return connection.play(createReadStream(location), { ...options, type: 'webm/opus' });
        }

        return connection.play(location, options);
    }

    async getTitle(location: string): Promise<string | null> {
        const { stdout } = await shell(`ffprobe -v quiet -print_format json -show_format -show_streams "${location}"`);
        logger.debug({ process: 'ffprobe', input: location, stdout });

        return JSON.parse(stdout).format?.tags?.title || null;
    }

    supports(location: string): boolean {
        return existsSync(location);
    }
}
