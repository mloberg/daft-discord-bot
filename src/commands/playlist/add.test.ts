import { Client, Message, TextChannel } from 'discord.js';
import { mocked } from 'ts-jest/utils';

import db from '../../db';
import { FriendlyError } from '../../error';
import logger from '../../logger';
import { hasPermission } from '../../permission';
import player from '../../player';
import command from './add';

const mocks = {
    react: jest.fn(),
    logInfo: mocked(logger.info),
    player: mocked(player),
    permission: mocked(hasPermission),
};

jest.mock('discord.js', () => ({
    Message: jest.fn().mockImplementation(() => ({
        member: { guild: { id: 'test', name: 'Test Guild' } },
        react: mocks.react,
    })),
}));

jest.mock('../../logger');
jest.mock('../../player');
jest.mock('../../permission');

describe('_playlist add', () => {
    let message: Message;

    beforeEach(async () => {
        mocks.react.mockClear();
        mocks.logInfo.mockClear();
        mocks.player.supports.mockClear();
        mocks.permission.mockClear();

        const client = {} as Client;
        const channel = {} as TextChannel;
        message = new Message(client, {}, channel);

        await db.playlist.create({ data: { name: 'test', guild: 'test', songs: ['test.mp3'] } });
        await db.playlist.create({ data: { name: 'empty', guild: 'test', songs: [] } });
        await db.playlist.create({ data: { name: 'foo', guild: 'another', songs: [] } });
    });

    afterEach(async () => {
        await db.$executeRaw`DELETE FROM playlists`;
        await db.$disconnect();
    });

    it('is a command', () => {
        expect(command.name).toBe('playlist add');
        expect(command).toMatchSnapshot();
    });

    it('adds songs to the end of a playlist', async () => {
        mocks.permission.mockReturnValue(true);
        mocks.player.supports.mockReturnValue(true);

        await command.run(message, { _: ['test', 'foo.mp3', 'bar.mp3'], $0: 'playlist add' });
        expect(mocks.react).toHaveBeenCalledWith('ðŸŽµ');
        expect(mocks.player.supports).toHaveBeenCalledTimes(2);
        expect(mocks.logInfo).toHaveBeenCalledWith('Added songs to playlist test in Test Guild');

        const playlist = await db.playlist.findFirst({ where: { name: 'test' } });
        expect(playlist).toMatchObject({
            name: 'test',
            guild: 'test',
            songs: ['test.mp3', 'foo.mp3', 'bar.mp3'],
        });
    });

    it('adds songs to empty playlists', async () => {
        mocks.permission.mockReturnValue(true);
        mocks.player.supports.mockReturnValue(true);

        await command.run(message, { _: ['empty', 'foo.mp3'], $0: 'playlist append' });
        expect(mocks.react).toHaveBeenCalledWith('ðŸŽµ');

        const playlist = await db.playlist.findFirst({ where: { name: 'empty' } });
        expect(playlist).toMatchObject({
            name: 'empty',
            guild: 'test',
            songs: ['foo.mp3'],
        });
    });

    it('throws an error if no name is given', async () => {
        mocks.permission.mockReturnValue(true);
        mocks.player.supports.mockReturnValue(true);

        await expect(command.run(message, { _: [], $0: 'playlist add' })).rejects.toThrow(
            new FriendlyError('I need a name for the playlist.'),
        );
    });

    it('throws an error if invalid songs are given', async () => {
        mocks.permission.mockReturnValue(true);
        mocks.player.supports.mockReturnValue(false);

        await expect(command.run(message, { _: ['test', 'foo.mp3'], $0: 'playlist add' })).rejects.toThrow(
            new FriendlyError('Cannot add to playlist. These songs are unsupported: foo.mp3'),
        );
    });

    it('throws an error if no playlist exists', async () => {
        mocks.permission.mockReturnValue(true);
        mocks.player.supports.mockReturnValue(true);

        await expect(command.run(message, { _: ['none', 'test.mp3'], $0: 'playlist append' })).rejects.toThrow(
            new FriendlyError("I couldn't find that playlist."),
        );
    });

    it('will not add songs to a playlist from another guild', async () => {
        mocks.permission.mockReturnValue(true);
        mocks.player.supports.mockReturnValue(true);

        await expect(command.run(message, { _: ['foo', 'test.mp3'], $0: 'playlist append' })).rejects.toThrow(
            new FriendlyError("I couldn't find that playlist."),
        );
    });

    it('will throw an error if user does not have role', async () => {
        mocks.permission.mockReturnValue(false);

        await expect(command.run(message, { _: ['test'], $0: 'playlist add' })).rejects.toThrow(
            new FriendlyError('You do not have permission to do that.'),
        );
    });
});
