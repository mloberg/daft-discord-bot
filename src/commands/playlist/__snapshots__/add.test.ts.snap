// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_playlist add is a command 1`] = `
Object {
  "alias": Array [
    "playlist append",
  ],
  "description": "Append songs to a playlist",
  "name": "playlist add",
  "run": [Function],
  "usage": "<name> [...songs]",
}
`;
